File: CanFinish.java
import java.util.Arrays;
import java.util.Queue;

/**
 * @author Vishal Puri
 * // Time Complexity : O(v+e)
 * // Space Complexity : O(v+e)
 * // Did this code successfully run on Leetcode : Yes
 * // Any problem you faced while coding this :
 */

public class CanFinish {
    public boolean canFinish(int numCourses, int[][] edges) {
        if (edges == null || edges.length == 0)
            return true;
        int indeg[] = new int[numCourses];
        Arrays.fill(indeg, 0);
        Queue<Integer> q = (Queue<Integer>) new LinkedList();
        int e = edges.length;
        for (int i = 0; i < e; i++) {
            if (edges[i][0] == edges[i][1])
                return false;
            indeg[edges[i][1]]++;
        }
        int count = 0;
        for (int i = 0; i < numCourses; i++) {
            if (indeg[i] == 0)
                q.add(i);
        }

        while (!q.isEmpty()) {
            int top = q.peek();
            System.out.println(top);
            q.remove();
            count++;
            for (int i = 0; i < e; i++) {

                if (edges[i][0] == top) {
                    indeg[edges[i][1]]--;
                    if (indeg[edges[i][1]] == 0)
                        q.add(edges[i][1]);
                }
            }
        }

        return count >= numCourses - 1 ? true : false;
    }
}

File: canFinish.swift


// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :


    func canFinish(_ numCourses: Int, _ prerequisites: [[Int]]) -> Bool {
        if numCourses == 0 || prerequisites.count == 0 || prerequisites == nil {
            return true
        }
        var inDegree = Array(repeating: 0, count: numCourses)
        var adjacencyList = [Int:[Int]]()
        var queue = [Int]()
        for course in prerequisites {
            var courseFrom = course[1]
            var courseTo = course[0]
            inDegree[courseTo] += 1
            adjacencyList[courseFrom,default: []].append(courseTo)
        }
        for i in 0..<inDegree.count {
            if inDegree[i] == 0 {
                queue.append(i)
            }
        }
        while !queue.isEmpty {
            var courseFrom = queue.removeFirst()
            if let list = adjacencyList[courseFrom] {
                for courseno in list {
                    inDegree[courseno] -= 1
                    if inDegree[courseno] == 0 {
                        queue.append(courseno)
                    }
                }
            }
        }
        for value in inDegree {
            print(value)
            if value != 0 {
                return false
            }
        }
        return true
    }
